src/loops.rs

```diff
@@ -228,6 +231,17 @@ impl LateLintPass for LoopsPass {
                 }
             }
         }
+        if let ExprMatch(ref expr, ref arms, MatchSource::WhileLetDesugar) = expr.node {
+            let pat = &arms[0].pats[0].node;
+            if let (&PatEnum(ref path, _), &ExprMethodCall(method_name, _, _)) = (pat, &expr.node) {
+                if method_name.node.as_str() == "next" &&
+                        match_trait_method(cx, expr, &["core", "iter", "Iterator"]) &&
+                        path.segments.last().unwrap().identifier.name.as_str() == "Some" {
+                    span_lint(cx, WHILE_LET_ON_ITERATOR, expr.span,
+                              "this loop could be written as a `for` loop");
```

Comment:
An obvious extension would be to extract the relevant parts of the code and rewrite the loop as a `for` loop using snippet.